[clang][scan-build] Treat --use-cc and --use-c++ as shell commands

rafl opened the pull request:
So that things like --use-cc="ccache gcc" work.

Fixes #26594.

Also use the slightly simpler shellwords instead of quotewords.

rafl committed:
[clang][scan-build] Treat --use-cc and --use-c++ as shell commands

So that things like --use-cc="ccache gcc" work.

Fixes #26594.

Also use the slightly simpler shellwords instead of quotewords.

balazs-benics-sonarsource commented:
Hey, thanks for coming by. Disclaimer: I'm not familiar with Perl.

When I looked at this patch and I had the impression that `CCC_CXX` is basically what one should use as `CXX` or `CMAKE_CXX_COMPILER`, which is just a path to a binary - without any arguments.
This patch would turn it into a combination of what other tools interpret as `CXX` and `CXX_FLAGS`, did I get this right?
This would mean we would deviate from the conventions of other tools, making it less ergonomic.

If we don't currently have a way to pass extra `CXX` flags, we should consider adding something like that instead of hacking the `CCC_CXX` option. WDYT?

rafl commented:
Thanks for your feedback, Balázs!

> When I looked at this patch and I had the impression that `CCC_CXX` is basically what one should use as `CXX` or `CMAKE_CXX_COMPILER`, which is just a path to a binary - without any arguments.

The author of the original code this change is modifying seemed to make the same assumption, but I believe this is not true. All of the following treats CC/CXX as a command with potential arguments:

```
CC="ccache gcc" ./configure
make CC="ccache gcc"
CXX="ccache g++" cmake ...
```

`CMAKE_<LANG>_COMPILER` also [supports potential arguments](https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_COMPILER.html):

```
cmake ... -DCMAKE_C_COMPILER='qcc;--arg1;--arg2'
```

The fact that all of these tools will consult `PATH` to resolve the compiler executable when it's not a fully-qualified path further supports that it's a command as opposed to a path to an executable.

> This patch would turn it into a combination of what other tools interpret as `CXX` and `CXX_FLAGS`, did I get this right? This would mean we would deviate from the conventions of other tools, making it less ergonomic.

Your understanding of what the proposed change does is correct. However, most if not all common build tools I've worked with treat these kinds of options as commands, and that's what users seem to expect as well, as per #26594.

This behaviour is very ergonomic, especially when you're truly swapping out the "compiler", such as when replacing "gcc" with "ccache gcc" or similar, without having to write wrapper executables, which is what's currently required by `scan-build` in these situations. In many situations it's possible to achieve the same goal using a combination of `CC` and `CFLAGS`, but this is often less desirable as it can quickly run into issues with accidentally overwriting `CFLAGS` from other sources or having to merge multiple `CFLAGS` correctly.

> If we don't currently have a way to pass extra `CXX` flags, we should consider adding something like that instead of hacking the `CCC_CXX` option. WDYT?

To me, following the established convention of `CC`/`CXX` being commands as opposed to paths to executables seems like the most straightforward approach, but I'm certainly open to alternatives if you see any practical downsides to this.

balazs-benics-sonarsource reviewed (approved):
Thanks for the context. It looks good to me now.
@haoNoQ, maybe you know some Perl, could you have a second opinion?

Otherwise, let's merge this in a week.

rafl commented:
Thanks for having another look @balazs-benics-sonarsource.

While we wait for a potential second opinion, I was curious about what you think of the current behaviour of silently defaulting to `$Default{CC,CXX}Compiler` if `CCC_{CC,CXX}` isn't executable and can't be resolved to an executable within `PATH`. I originally found this surprising when `CC="ccache gcc"` would be effectively ignored. Would it be worth changing this while we're at it?

Would you have any preference between these options?

1. keeping it as it is
2. warning when ignoring an option and falling back to a default
3. providing an obvious error

My personal preference is option 3. The only downside to this which I see is potentially breaking some existing usage, but it'd do so in an obvious and easy to fix way way. I'm not sure if that's acceptable in the context of this project.

Option 2 would be my second choice, with Option 1 being the least preferable.

balazs-benics-sonarsource commented:
I'd prefer option 2, because why else would we have a default compiler if that wasn't used in some workflows.
A warning could never hurt. I'm also flexible on the subject.

rafl committed:
[clang][scan-build] Minor simplification/refactor of environment detection

I want to adjust the logic for CCC_CC and CCC_CXX defaults, so not
repeating that logic twice is convenient.

As a nice side-effect, we end up with fewer globals and less code.

Contains no behavioural changes, except for checking for /usr/bin/xcrun
even if we're not on OSX when there's an -isysroot parameter, which
should be fine. We also don't call uname -s twice.

rafl committed:
[clang][scan-build] Warn about unusable CCC_CC and CCC_CXX option values

To avoid user confusion when we're ignoring the provided option and use
a default instead.

rafl committed:
[clang][scan-build] Trust users' compiler choice

If you went out of your way to specify the option, you probably knew
what you wanted and didn't want the default you'd get from not
specifying the option in the first place.

We just let the error from from system() propagate to tell the user that
the option they provided wasn't usable:

  $ scan-build-19 --use-cc="enoent gcc" make
  scan-build: Using '/usr/lib/llvm-19/bin/clang' for static analysis
  ...
  Can't exec "enoent": No such file or directory ...

rafl committed:
[clang][scan-build] Mirror CCC_{CC,CXX} behaviour for CLANG{,_CXX}

rafl commented:
> I'd prefer option 2, because why else would we have a default compiler if that wasn't used in some workflows. A warning could never hurt.

`--use-cc`/`--use-c++`/`CCC_CC`/`CCC_CXX` are optional, so the default would still be used when those options are not specified, which might be the majority of use-cases.

I've added a few additional commits:

* 6806e42d3c76 includes some minor simplifications and deduplicates the logic we're talking about without changing behaviour.
*  e101c97b60f2 implements option 2, as that was your preference.
*  b086d3ecad7e implements my preferred option 3. I'd happy to drop this commit if we wanna stick with option 2 instead.
* 805396fa1ec1 brings the handling of `CLANG`/`CLANG_CXX` in line with `CCC_CC`/`CCC_CXX` if we end up going for option 3. We'd also drop this if we go for option 2.

balazs-benics-sonarsource commented:
Looks good as it is right now. Thanks for putting the effort into this.
I've invited the rest of the folks probably interested in this to get a second opinion.

NagyDonat commented on clang/tools/scan-build/libexec/ccc-analyzer:
```diff
@@ -51,63 +51,40 @@ sub silent_system {
 
-  $IsCXX = 0
+sub DetermineClang {
+  my ($is_cxx) = @_;
```
Bikeshedding: Please use `$UpperCamelCase` within the new functions `DetermineCompiler` / `DetermineClang` because it's a bit jarring when `$IsCXX` / `$is_cxx` appears both in `$UpperCamelCase` and in `$snake_case`. I see that these styles both appear in the script, but in these parts  `$UpperCamelCase` is more common, so I think it would be better if you used that.

rafl replied:
I'm with you on the mix of different styles in this code being far from ideal. I did briefly consider cleaning those things up a bit, but decided to not extend the scope of this changeset further and to perhaps save those kinds of changes for a separate pull-request if there was any interest in that.

Here's how I arrived at the names currently used in this pull-request:

* Functions
Almost all existing functions use UpperCamelCase, so I stuck with that even though it's not at all typical for the majority of modern(-ish) Perl code
* Variables
These were also quite inconsistent, but I stuck with common modern(-ish) practices, where most regular lexical variables use `$snake_case`, constants use `$ALL_CAPS`, and "globals" use `$UpperCamelCase` as per https://metacpan.org/dist/perl/view/pod/perlstyle.pod, the old Perl Best Practices book, and established community conventions.

  I'd find it much more jarring if a function was to potentially shadow a variable which reads like a `$Global` from a surrounding scope.

I don't feel particularly strongly about it, and would appreciate your input on the following suggestions for how to best move forwards:

1. Just call the parameter `$IsCXX` in the functions as suggested.
This defies the expectations of many Perl programmers when it comes to naming, but so does most of the rest of the program.
2. Keep things as they are.
This is consistent with modern practices.
3. Call the local `$is_cxx` variable something else (but not `$IsCXX`), or rename the "global" `$IsCXX`.
To not potentially confuse people unfamiliar with Perl or its current common practices.
4. Just use the "global" `$IsCXX` directly.
Reducing some of the global state to make it slightly easier to reason about those individual functions seemed nice, but given how much of it there is anyway, maybe it doesn't even matter in this context.

Thanks for your review and your feedback on the above!

NagyDonat replied:
Ok, thanks for the clarifications. I didn't know about these Perl naming conventions (last time I touched Perl was 10+ years ago), but now that you explained them, I agree that it's reasonable to follow them.

NagyDonat reviewed (approved):
Overall I'm satisfied with the direction of the commit and I agree that explicitly specifying an invalid compiler should be a hard error. (I could also accept a warning as long as it's loud and visible, but the current "silently use the default" behavior is terrible and must be replaced.)

ziqingluo-90 commented on clang/tools/scan-build/libexec/ccc-analyzer:
```diff
@@ -51,63 +51,40 @@ sub silent_system {
 
-$AnalyzerTarget = $ENV{'CLANG_ANALYZER_TARGET'};
+my $IsCXX = $FindBin::Script =~ /c\+\+-analyzer/;
+my ($Compiler, @CompilerArgs) = DetermineCompiler($IsCXX);
```
Hi @rafl .
I'm not familiar with Perl.  Just curious, how does it parse, for example, `ccache gcc`?  Would `$Compiler` hold `ccache` and `gcc` becomes one of the `@CompilerArgs`?

rafl replied:
@ziqingluo-90 that's exactly right. `DetermineCompiler` returns a list of values (potentially a singleton list), and the call-site unpacks that return value into the first list element `$Compiler` and the rest of the list `@CompilerArgs` (which might be empty).

The actual parsing is done using shellwords from [Text::ParseWords](https://metacpan.org/pod/Text::ParseWords).

If you'd like to experiment how it parses certain inputs, you can do something like this, which prints out the parsed list, one item per line:

```
$ CCC_CC='foo "bar baz" \"quux' perl -MText::ParseWords -le'print for shellwords $ENV{CCC_CC}'
foo
bar baz
"quux
```
